pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -1349,21 +1370,13 @@ char **NpyArr_encodeLabels(PyArrayObject *labels, PyObjectEncoder *enc,
             }
         } else {
             // NA values need special handling
-            if (PyFloat_Check(item)) {
-                double fval = PyFloat_AS_DOUBLE(item);
-                is_null = npy_isnan(fval);
-            } else if (item == Py_None || object_is_na_type(item)) {
-                is_null = 1;
-            } else if (object_is_decimal_type(item)) {
-                PyObject *is_null_obj = PyObject_CallMethod(item,
-                                                            "is_nan",
-                                                            NULL);
-                is_null = (is_null_obj == Py_True);
-                if (!is_null_obj) {
-                    goto INVALID;
-                }
-                Py_DECREF(is_null);
-            } else {
+            is_null = is_null_obj(item);
+            if (is_null == -1) {
+                // Something errored
+                // Return to let the error surface
+                return 0;
```

Comment:
Surprised we return 0 here but I see that you are just matching the pattern of the rest of the function. It should really be returning NULL in case of an error - looks to not be handled properly. But of course that is separate from this PR